cirq/optimizers/controlled_gate_decomposer.py

```diff
@@ -0,0 +1,284 @@
+
+        Returns sequence of operations, which, when applied in sequence, gives
+        result equivalent to applying single-qubit gate with matrix
+        `matrix^power` on `target`, controlled by `controls`.
+
+        Result is guaranteed to consist exclusively of 1-qubit, CNOT and CCNOT
+        gates.
+
+        If matrix is special unitary, result has length `O(len(controls))`.
+        Otherwise result has length `O(len(controls)**2)`.
+
+        Args:
+            matrix - 2x2 numpy unitary matrix (of real or complex dtype).
+            controls - control qubits.
+            targets - target qubits.
+            free_qubits - qubits which are neither controlled nor target. Can
+                be modified by algorithm but will end up in inital state.
+            power - power in which `unitary` should be raised. Deafults to 1.0.
+        """
+        if free_qubits is None:
+            free_qubits = []
+        assert linalg.is_unitary(matrix)
+        assert matrix.shape == (2, 2)
+        u = _unitary_power(matrix, power)
+
+        # Matrix parameters, see definitions in [1], chapter 4.
+        delta = np.angle(np.linalg.det(u)) * 0.5
+        alpha = np.angle(u[0, 0]) + np.angle(u[0, 1]) - 2 * delta
+        beta = np.angle(u[0, 0]) - np.angle(u[0, 1])
+        theta = 2 * np.arccos(np.minimum(1.0, np.abs(u[0, 0])))
+
+        # Decomposing matrix into three matrices - see [1], lemma 4.3.
+        A = _rz(alpha) @ _ry(theta / 2)
+        B = _ry(-theta / 2) @ _rz(-(alpha + beta) / 2)
+        C = _rz((beta - alpha) / 2)
+        X = np.array([[0, 1], [1, 0]])
+        assert np.allclose(A @ B @ C, np.eye(2), atol=1e-2)
+        assert np.allclose(A @ X @ B @ X @ C, u / np.exp(1j * delta), atol=1e-2)
+
+        m = len(controls)
+        ctrl = controls
+        assert m > 0, "No control qubits."
+
+        if m == 1:
+            # See [1], chapter 5.1.
+            result = [
+                ops.ZPowGate(exponent=delta / np.pi).on(ctrl[0]),
+                ops.rz(-0.5 * (beta - alpha)).on(target),
+                ops.CNOT.on(ctrl[0], target),
+                ops.rz(0.5 * (beta + alpha)).on(target),
+                ops.ry(0.5 * theta).on(target),
+                ops.CNOT.on(ctrl[0], target),
+                ops.ry(-0.5 * theta).on(target),
+                ops.rz(-alpha).on(target),
+            ]
+
+            # Remove no-ops.
+            result = [g for g in result if not _is_identity(g._unitary_())]
+
+            return result
+        else:
+            gate_is_special_unitary = np.allclose(delta, 0)
+
+            if gate_is_special_unitary:
+                # O(n) decomposition of SU matrix - [1], lemma 7.9.
+                cnot_seq = self.decompose_x(ctrl[:-1],
+                                            target,
+                                            free_qubits=[ctrl[-1]])
+                result = []
+                result += self.decompose(C, [ctrl[-1]], target)
```

Comment:
I would just use the operation `C(target).controlled_by(ctrl[-1])` and rely on the caller to further decompose if they need to.

Reply:
I am implementing an algorithm which results in arbitrary single-qubit gates, CNOT and CCNOT gates - arbitrary single-controlled gates are not allowed. That's my understanding of the original paper.

Maybe it makes more sense to not decompose single-controlled gates and return them "as is", but it will be a different algorithm.